### PR TITLE
implement KVS transaction object

### PR DESCRIPTION
**Pull request by garlick**
This PR implements the proposed `flux_kvs_txn_t` objects for the KVS discussed in #1094:
```c
flux_kvs_txn_t *flux_kvs_txn_create (void);
void flux_kvs_txn_destroy (flux_kvs_txn_t *txn);

int flux_kvs_txn_put (flux_kvs_txn_t *txn, int flags,
                      const char *key, const char *json_str);

int flux_kvs_txn_pack (flux_kvs_txn_t *txn, int flags,
                       const char *key, const char *fmt, ...);

int flux_kvs_txn_link (flux_kvs_txn_t *txn, int flags,
                       const char *key, const char *json_str);

int flux_kvs_txn_mkdir (flux_kvs_txn_t *txn, int flags,
                        const char *key);

int flux_kvs_txn_unlink (flux_kvs_txn_t *txn, int flags,
                         const char *key);

int flux_kvs_txn_symlink (flux_kvs_txn_t *txn, int flags,
                          const char *key, const char *target);

flux_future_t *flux_kvs_commit (flux_t *h, int flags, flux_kvs_txn_t *txn);

flux_future_t *flux_kvs_fence (flux_t *h, int flags, const char *name,
                               int nprocs, flux_kvs_txn_t *txn);

```
As discussed in #1094 we'll want some convenience functions since having to create/put/commit/destroy to change a single value is a bit cumbersome.  For now the old interfaces are still there, rewritten in terms of the above.

This PR also internally converts the KVS client API to use jansson.

### CODE REVIEW

**Comment by chu11** [comment]
At a high level, things look good to me.  I think it was a good idea to document the json "null" vs C NULL, which can be confusing.

**Comment by garlick** [comment]
Doing a bit of tidying up on this PR.  Hopefully should be ready for another look by mid morning, then if no objection's I'll squash down the incremental changes and ask for merge consideration.

**Comment by garlick** [comment]
OK, done with cleanup.  This is ready for review IMHO.

**Comment by garlick** [comment]
rebased and added one last test.

**Comment by garlick** [comment]
Had one travis builder stall here:
```
PASS: t2000-wreck.t 5 - wreckrun: propagates current working directory
PASS: t2000-wreck.t 6 - wreckrun: propagates current environment
*** glibc detected *** lua: corrupted double-linked list: 0x0000000000dc2590 ***
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```
Hmm.  Restarting.

**Comment by grondo** [comment]
Maybe it should be obvious, but perhaps the documentation for `flux_kvs_txn_link` could be improved.

> `flux_kvs_txn_link()` sets _key_ to be a hard link to a KVS tree object looked up with FLUX_KVS_TREEOBJ.

Perhaps two suggestions to make this clearer:

 * Replace the generic `const char *json_str` argument name in the prototype with `const char *treeobj` for symmetry with other calls that take treeobj JSON arguments.
 * Instead of referencing only the `FLUX_KVS_TREEOBJ` enum, would it be ok to reword the above liek

> `flux_kvs_txn_link()` sets _key_ to be a hard link to a KVS tree object obtained with `flux_kvs_lookup()` or `flux_kvs_lookupat()` using the `FLUX_KVS_TREEOBJ` flag.

(or similar)

Also, how does this function compare to the "classic" call `flux_kvs_put_treeobj`?

**Comment by garlick** [comment]
Those changes sound good, patch coming.

> Also, how does this function compare to the "classic" call flux_kvs_put_treeobj?

Identical except that `kvs_put_treeobj()` operates on the default transaction rather than an explict one.

**Comment by grondo** [comment]
> Identical except that kvs_put_treeobj() operates on the default transaction rather than an explict one.

Thanks. In a way the name of `kvs_put_treeobj` was more clear, so I was wondering about the change to simply "`link`" (sorry if this was already discussed and I forgot!). `flux_kvs_txn_link()` makes me think of the `link(2)` syscall so I was expecting the arguments to be similar -- e.g. `link (const char *oldpath, const char *newpath)`, which might have added to the confusion.

**Comment by grondo** [comment]
The only other question I'll pose is whether the `_txn_` in the function names was dropped purposefully for `flux_kvs_commit` and `flux_kvs_commit`. If there is any chance "implied transaction" versions of these functions would be needed in the future, we might want to reserve the non-`txn` names.

**Comment by grondo** [comment]
Other than those minor questions, I did  test out this PR on some of our TOSS2 and 3 systems and no gotchas so I'm willing to merge for you once you're ready.

**Comment by garlick** [comment]
There wasn't any discussion that I recall.  I sort of made that change on my own initiative, and I'm fine with changing it back.  I find my thinking on interface design is often a bit warped when I've had my head in the implementation details.

I'll go ahead and change that, and squash the incremental changes.  Thanks for the testing and review!

**Comment by garlick** [comment]
Actually, maybe for symmetry with `flux_kvs_lookup(key, flags=FLUX_KVS_TREEOBJ)`, we should drop the extra function and use a FLUX_KVS_TREEOBJ flag with `flux_kvs_txn_put()`?

**Comment by grondo** [comment]
> Actually, maybe for symmetry with flux_kvs_lookup(key, flags=FLUX_KVS_TREEOBJ), we should drop the extra function and use a FLUX_KVS_TREEOBJ flag with flux_kvs_txn_put()?

That works for me if it is OK with you!

**Comment by garlick** [comment]
Just made the discussed change, squashed incremental development, and forced a push.

**Comment by garlick** [comment]
Uh oh, one builder hung here again:
```
PASS: t2002-pmi.t 8 - pmi: wreck sets PMI_FD
PASS: t2002-pmi.t 9 - pmi: wreck preputs PMI_process_mapping into kvs
PASS: t2002-pmi.t 10 - pmi: PMI reports correct rank, size
*** glibc detected *** lua: corrupted double-linked list: 0x0000000001ce49e0 ***
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```
well, same error, different output.  I'll restart.  Not sure if this is a real problem or travis weirdness..

**Comment by grondo** [comment]
I didn't see this once running make check on other systems, so I'm not sure what is going on here :frowning_face:

Maybe we should try running make check in a loop on a system from which we could at least get a corefile or something...

**Comment by garlick** [comment]
OK, I can start that up on my desktop...

File: src/common/libkvs/jansson_dirent.c [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+    if (!strcmp (type, "FILEREF") || !strcmp (type, "DIRREF")) {
+        char *ref = arg;
+        json_t *o;
+
+        if (!(o = json_string (ref)))
+            oom ();
+        if (json_object_set_new (dirent, type, o) < 0)
+            oom ();
+
```
Comment (grondo) [inline src/common/libkvs/jansson_dirent.c]: Just noticed this, is it OK we're using `oom()` here in newly created sources?

**Comment by grondo** [comment]
Also, code coverage seems to indicate "classic" `kvs_put_double` and `kvs_put_boolean` calls are not used? (I had thought we'd got rid of some of those, but perhaps those were get variants?)

**Comment by garlick** [comment]
Ah yes I could probably get rid of those unused interfaces.  Sorry, forgot to do that!

Regarding the oom calls in jansson_dirent(), the new API calls are only calling `j_dirent_validate()` which doesn't oom.  The kvs module does use it in PR #1108, but only on the road towards the treeobj conversion, when this particular code goes away.  @chu11 and I discussed that verbally and thought it was ok to avoid complicating the code that uses it while it's in transition.

Just got a failure int the third iteration of my test loop:
```
FAIL: t2002-pmi.t 4 - pmi: wreck sets FLUX_LOCAL_RANKS multiple tasks per node
#       
#               run_program 5 ${SIZE} 1 printenv FLUX_LOCAL_RANKS >output_clique &&
#               test `cut -d: -f2 output_clique | uniq` = `seq -s, 0 $((${SIZE}-1))`
#       

```
not a hang though and no core...

File: src/common/libkvs/jansson_dirent.c (hunk 1 above)
Comment (chu11) [inline src/common/libkvs/jansson_dirent.c]: Perhaps @garlick and I talked about this offline.  The "classic" dirent functions (```dirent_create()```, etc.) all used ```oom()``` behind the scenes.  I think refactoring the jansson equivalents (```j_dirent_create()```, etc.) to return ENOMEM and all callers to handle this is probably a bit much refactoring for this round.   Figured to leave the ```oom()``` behavior and leave that refactoring for another round.

**Comment by grondo** [comment]
Thanks guys, just wanted verification.

**Comment by garlick** [comment]
Running t2002-pmi.t in a loop I was finally able to get a core, which points to my changes in the kvs_watch code.  Glad we ran that down!  Will dig into this tomorrow.
```
$ gdb /usr/bin/lua  ./t/trash-directory.t2002-pmi/core
[snip]
(gdb) bt full
(gdb) bt full
#0  _int_malloc (av=av@entry=0x7f8e62bfbb20 <main_arena>, 
    bytes=bytes@entry=128) at malloc.c:3516
        iters = <optimized out>
        nb = 144
        idx = 9
        bin = <optimized out>
        victim = 0x1709fef
        size = 20736
        victim_index = <optimized out>
        remainder = <optimized out>
        remainder_size = <optimized out>
        block = <optimized out>
        bit = <optimized out>
        map = <optimized out>
        fwd = <optimized out>
        bck = 0x1703e3000
        errstr = 0x0
        __func__ = "_int_malloc"
#1  0x00007f8e628bb184 in __GI___libc_malloc (bytes=128) at malloc.c:2913
        ar_ptr = 0x7f8e62bfbb20 <main_arena>
        victim = <optimized out>
        hook = <optimized out>
#2  0x00007f8e621a813d in ?? () from /usr/lib/x86_64-linux-gnu/libjansson.so.4
No symbol table info available.
#3  0x00007f8e621ac7e7 in json_object ()
   from /usr/lib/x86_64-linux-gnu/libjansson.so.4
No symbol table info available.
#4  0x00007f8e621a98e5 in ?? () from /usr/lib/x86_64-linux-gnu/libjansson.so.4
No symbol table info available.
#5  0x00007f8e621a999d in ?? () from /usr/lib/x86_64-linux-gnu/libjansson.so.4
No symbol table info available.
#6  0x00007f8e621a999d in ?? () from /usr/lib/x86_64-linux-gnu/libjansson.so.4
No symbol table info available.
#7  0x00007f8e621a9dae in ?? () from /usr/lib/x86_64-linux-gnu/libjansson.so.4
No symbol table info available.
#8  0x00007f8e621a9f36 in json_loads ()
   from /usr/lib/x86_64-linux-gnu/libjansson.so.4
No symbol table info available.
#9  0x00007f8e61b2e798 in flux_msg_vunpack (cmsg=0x170e910, 
    fmt=<optimized out>, ap=ap@entry=0x7ffe8a893ed0) at message.c:1171
        rc = -1
        json_str = 0x170d3a8 "{ \"val\": { \"000000\": { \"FILEVAL\": { \"eof\": true, \"data\": \"\" } } } }"
        error = {line = -1, column = -1, position = 0, 
          source = "<string>", '\000' <repeats 28 times>, "\001\000\000\000\004\000\000\000\300?p\001\000\000\000\000\000\000p\001", '\000' <repeats 12 times>, "\300?\211\212\376\177\000\000\001\000\000", 
          text = "\000\000\000\000\377\377\377\377\377\377\377\377\000\000\000\000\000\000\000\000X\273\277b\216\177\000\000`>\211\212\376\177\000\000\377\377\377\377\216\177\000\000\241\301vu\001\200\377\377_>\211\212\376\177\000\000\240\244\277b\216\177\000\000\000\000\000\000\000\000\000\000p\237p\001\000\000\000\000\300?\211\212\000\000\000\000\031", '\000' <repeats 15 times>, "@>p\001\000\000\000\000(\240p\001\000\000\000\000P\353p\001\000\000\000\000\300?\211\212\376\177\000\000\360\001q\001\000\000\000\000 \273\277b\216\177\000\000\020\351p\001"}
        msg = 0x170e910
#10 0x00007f8e61b2e894 in flux_msg_unpack (msg=msg@entry=0x170e910, 
    fmt=fmt@entry=0x7f8e61b64772 "{s:o}") at message.c:1191
        ap = <error reading variable ap (Attempt to dereference a generic pointer.)>
        rc = <optimized out>
#11 0x00007f8e61b371c4 in watch_response_cb (h=0x16b5bf0, w=<optimized out>, 
    msg=0x170e910, arg=<optimized out>) at kvs_watch.c:217
        json_str = 0x0
        val = 0x0
        matchtag = 26
#12 0x00007f8e61b29ef4 in call_handler (w=w@entry=0x16b8fe0, 
    msg=msg@entry=0x170e910) at dispatch.c:302
        rolemask = 1
        matchtag = 32766
#13 0x00007f8e61b2a09a in dispatch_message (type=2, msg=0x170e910, d=0x1691070)
    at dispatch.c:325
        w = 0x16b8fe0
        match = <optimized out>
#14 handle_cb (r=0x16667b0, hw=<optimized out>, revents=<optimized out>, 
    arg=0x1691070) at dispatch.c:391
        d = 0x1691070
        msg = 0x170e910
        rc = -1
        type = 2
        topic = 0x1708590 "kvs.watch"
        __FUNCTION__ = "handle_cb"
#15 0x00007f8e61b4e2b3 in ev_invoke_pending (loop=0x16a00c0) at ev.c:3314
        p = <optimized out>
#16 0x00007f8e61b5190e in ev_run (loop=0x16a00c0, flags=0) at ev.c:3717
        flags = 0
        loop = 0x16a00c0
#17 0x00007f8e61b293a3 in flux_reactor_run (r=0x16667b0, flags=flags@entry=0)
    at reactor.c:128
        ev_flags = 0
        count = <optimized out>
#18 0x00007f8e623c812d in l_flux_reactor_start (L=0x1650010) at flux-lua.c:1895
        rc = <optimized out>
        arg = <optimized out>
        reason = <optimized out>
        mode = 0
#19 0x00000000004080e0 in ?? ()
No symbol table info available.
#20 0x00000000004116e2 in ?? ()
No symbol table info available.
#21 0x000000000040853d in ?? ()
No symbol table info available.
#22 0x000000000040782b in ?? ()
No symbol table info available.
#23 0x000000000040869b in ?? ()
No symbol table info available.
#24 0x0000000000405d26 in lua_pcall ()
No symbol table info available.
#25 0x0000000000403ecc in ?? ()
No symbol table info available.
#26 0x00000000004048a9 in ?? ()
No symbol table info available.
#27 0x00000000004080e0 in ?? ()
No symbol table info available.
#28 0x00000000004084f4 in ?? ()
No symbol table info available.
#29 0x000000000040782b in ?? ()
No symbol table info available.
#30 0x000000000040869b in ?? ()
No symbol table info available.
#31 0x0000000000405db5 in lua_cpcall ()
No symbol table info available.
#32 0x0000000000403b94 in main ()
No symbol table info available.
```

**Comment by garlick** [comment]
Same backtrace with libjansson-dbg installed FWIW:
```
(gdb) bt full
#0  _int_malloc (av=av@entry=0x7f8e62bfbb20 <main_arena>, 
    bytes=bytes@entry=128) at malloc.c:3516
        iters = <optimized out>
        nb = 144
        idx = 9
        bin = <optimized out>
        victim = 0x1709fef
        size = 20736
        victim_index = <optimized out>
        remainder = <optimized out>
        remainder_size = <optimized out>
        block = <optimized out>
        bit = <optimized out>
        map = <optimized out>
        fwd = <optimized out>
        bck = 0x1703e3000
        errstr = 0x0
        __func__ = "_int_malloc"
#1  0x00007f8e628bb184 in __GI___libc_malloc (bytes=128) at malloc.c:2913
        ar_ptr = 0x7f8e62bfbb20 <main_arena>
        victim = <optimized out>
        hook = <optimized out>
#2  0x00007f8e621a813d in hashtable_init (hashtable=hashtable@entry=0x1704850)
    at hashtable.c:188
        i = <optimized out>
#3  0x00007f8e621ac7e7 in json_object () at value.c:64
        object = <optimized out>
#4  0x00007f8e621a98e5 in parse_object (error=0x7ffe8a893da0, flags=0, 
    lex=0x7ffe8a893ce0) at load.c:681
        object = <optimized out>
#5  parse_value (lex=lex@entry=0x7ffe8a893ce0, flags=flags@entry=0, 
    error=error@entry=0x7ffe8a893da0) at load.c:854
        json = <optimized out>
        value = 0
#6  0x00007f8e621a999d in parse_object (error=0x7ffe8a893da0, flags=0, 
    lex=0x7ffe8a893ce0) at load.c:724
        key = 0x1703fa0 "000000"
        len = <optimized out>
        value = <optimized out>
        object = 0x1709610
#7  parse_value (lex=lex@entry=0x7ffe8a893ce0, flags=flags@entry=0, 
    error=error@entry=0x7ffe8a893da0) at load.c:854
        json = <optimized out>
        value = 0
#8  0x00007f8e621a999d in parse_object (error=0x7ffe8a893da0, flags=0, 
    lex=0x7ffe8a893ce0) at load.c:724
        key = 0x1708400 "val"
        len = <optimized out>
        value = <optimized out>
        object = 0x1703f30
#9  parse_value (lex=lex@entry=0x7ffe8a893ce0, flags=flags@entry=0, 
    error=error@entry=0x7ffe8a893da0) at load.c:854
        json = <optimized out>
        value = 2.0821863053081827e-317
#10 0x00007f8e621a9dae in parse_json (lex=lex@entry=0x7ffe8a893ce0, 
    flags=flags@entry=0, error=error@entry=0x7ffe8a893da0) at load.c:888
        result = <optimized out>
#11 0x00007f8e621a9f36 in json_loads (
    string=0x170d3a8 "{ \"val\": { \"000000\": { \"FILEVAL\": { \"eof\": true, \"data\": \"\" } } } }", flags=flags@entry=0, error=error@entry=0x7ffe8a893da0)
    at load.c:948
        lex = {stream = {get = 0x7f8e621a88d0 <string_get>, 
            data = 0x7ffe8a893cd0, buffer = "{\000\211\212\376", 
            buffer_pos = 1, state = 0, line = 1, column = 22, 
            last_column = 32654, position = 22}, saved_text = {
            value = 0x17085f0 "{", length = 1, size = 16}, token = 123, 
          value = {string = {val = 0x0, len = 0}, integer = 0, real = 0}}
        result = <optimized out>
        stream_data = {
          data = 0x170d3a8 "{ \"val\": { \"000000\": { \"FILEVAL\": { \"eof\": true, \"data\": \"\" } } } }", pos = 22}
#12 0x00007f8e61b2e798 in flux_msg_vunpack (cmsg=0x170e910, 
    fmt=<optimized out>, ap=ap@entry=0x7ffe8a893ed0) at message.c:1171
        rc = -1
        json_str = 0x170d3a8 "{ \"val\": { \"000000\": { \"FILEVAL\": { \"eof\": true, \"data\": \"\" } } } }"
        error = {line = -1, column = -1, position = 0, 
          source = "<string>", '\000' <repeats 28 times>, "\001\000\000\000\004\000\000\000\300?p\001\000\000\000\000\000\000p\001", '\000' <repeats 12 times>, "\300?\211\212\376\177\000\000\001\000\000", 
          text = "\000\000\000\000\377\377\377\377\377\377\377\377\000\000\000\000\000\000\000\000X\273\277b\216\177\000\000`>\211\212\376\177\000\000\377\377\377\377\216\177\000\000\241\301vu\001\200\377\377_>\211\212\376\177\000\000\240\244\277b\216\177\000\000\000\000\000\000\000\000\000\000p\237p\001\000\000\000\000\300?\211\212\000\000\000\000\031", '\000' <repeats 15 times>, "@>p\001\000\000---Type <return> to continue, or q <return> to quit---
\000\000(\240p\001\000\000\000\000P\353p\001\000\000\000\000\300?\211\212\376\177\000\000\360\001q\001\000\000\000\000 \273\277b\216\177\000\000\020\351p\001"}
        msg = 0x170e910
#13 0x00007f8e61b2e894 in flux_msg_unpack (msg=msg@entry=0x170e910, 
    fmt=fmt@entry=0x7f8e61b64772 "{s:o}") at message.c:1191
        ap = <error reading variable ap (Attempt to dereference a generic pointer.)>
        rc = <optimized out>
#14 0x00007f8e61b371c4 in watch_response_cb (h=0x16b5bf0, w=<optimized out>, 
    msg=0x170e910, arg=<optimized out>) at kvs_watch.c:217
        json_str = 0x0
        val = 0x0
        matchtag = 26
#15 0x00007f8e61b29ef4 in call_handler (w=w@entry=0x16b8fe0, 
    msg=msg@entry=0x170e910) at dispatch.c:302
        rolemask = 1
        matchtag = 32766
#16 0x00007f8e61b2a09a in dispatch_message (type=2, msg=0x170e910, d=0x1691070)
    at dispatch.c:325
        w = 0x16b8fe0
        match = <optimized out>
#17 handle_cb (r=0x16667b0, hw=<optimized out>, revents=<optimized out>, 
    arg=0x1691070) at dispatch.c:391
        d = 0x1691070
        msg = 0x170e910
        rc = -1
        type = 2
        topic = 0x1708590 "kvs.watch"
        __FUNCTION__ = "handle_cb"
#18 0x00007f8e61b4e2b3 in ev_invoke_pending (loop=0x16a00c0) at ev.c:3314
        p = <optimized out>
#19 0x00007f8e61b5190e in ev_run (loop=0x16a00c0, flags=0) at ev.c:3717
        flags = 0
        loop = 0x16a00c0
#20 0x00007f8e61b293a3 in flux_reactor_run (r=0x16667b0, flags=flags@entry=0)
    at reactor.c:128
        ev_flags = 0
        count = <optimized out>
#21 0x00007f8e623c812d in l_flux_reactor_start (L=0x1650010) at flux-lua.c:1895
        rc = <optimized out>
        arg = <optimized out>
        reason = <optimized out>
        mode = 0
```

**Comment by grondo** [comment]
Even with the above fix (which I think is valid, good catch), there are some other issues being caught by running  the lua test `t/lua/t1002-kvs.t` under valgrind (by modification of fluxometer.lua). Possibly there are more places where the lua bindings should be taking an extra reference on returned objects, though I can't make sense of these errors yet:

```
==96418== Invalid read of size 8
==96418==    at 0x7436815: ??? (in /usr/lib64/libjansson.so.4.9.0)
==96418==    by 0x74368B8: ??? (in /usr/lib64/libjansson.so.4.9.0)
==96418==    by 0x743B7F8: json_delete (in /usr/lib64/libjansson.so.4.9.0)
==96418==    by 0x6FDBE7C: json_decref (jansson.h:112)
==96418==    by 0x6FDBE7C: flux_msg_destroy (message.c:263)
==96418==    by 0x6FD706D: handle_cb (dispatch.c:445)
==96418==    by 0x6FFBD3C: ev_invoke_pending (ev.c:3314)
==96418==    by 0x7000369: ev_run (ev.c:3717)
==96418==    by 0x6FD6D2E: flux_reactor_run (reactor.c:128)
==96418==    by 0x6D871A8: l_flux_reactor_start (flux-lua.c:1895)
==96418==    by 0x4E3E5A0: ??? (in /usr/lib64/liblua-5.1.so)
==96418==    by 0x4E49228: ??? (in /usr/lib64/liblua-5.1.so)
==96418==    by 0x4E3EA6C: ??? (in /usr/lib64/liblua-5.1.so)
==96418==  Address 0x6213618 is 8 bytes inside a block of size 32 free'd
==96418==    at 0x4C28C64: free (vg_replace_malloc.c:530)
==96418==    by 0x6FE4BEC: json_decref (jansson.h:112)
==96418==    by 0x6FE4BEC: watch_response_cb (kvs_watch.c:227)
==96418==    by 0x6FD6FC3: call_handler (dispatch.c:302)
==96418==    by 0x6FD719E: dispatch_message (dispatch.c:325)
==96418==    by 0x6FD719E: handle_cb (dispatch.c:391)
==96418==    by 0x6FFBD3C: ev_invoke_pending (ev.c:3314)
==96418==    by 0x7000369: ev_run (ev.c:3717)
==96418==    by 0x6FD6D2E: flux_reactor_run (reactor.c:128)
==96418==    by 0x6D871A8: l_flux_reactor_start (flux-lua.c:1895)
==96418==    by 0x4E3E5A0: ??? (in /usr/lib64/liblua-5.1.so)
==96418==    by 0x4E49228: ??? (in /usr/lib64/liblua-5.1.so)
==96418==    by 0x4E3EA6C: ??? (in /usr/lib64/liblua-5.1.so)
==96418==    by 0x4E3E106: ??? (in /usr/lib64/liblua-5.1.so)
==96418==  Block was alloc'd at
==96418==    at 0x4C2928A: malloc (vg_replace_malloc.c:299)
==96418==    by 0x743B0AD: ??? (in /usr/lib64/libjansson.so.4.9.0)
==96418==    by 0x743803A: ??? (in /usr/lib64/libjansson.so.4.9.0)
==96418==    by 0x7437E46: ??? (in /usr/lib64/libjansson.so.4.9.0)
==96418==    by 0x743820C: ??? (in /usr/lib64/libjansson.so.4.9.0)
==96418==    by 0x7438729: json_loads (in /usr/lib64/libjansson.so.4.9.0)
==96418==    by 0x6FDB731: flux_msg_vunpack (message.c:1171)
==96418==    by 0x6FDB7F7: flux_msg_unpack (message.c:1191)
==96418==    by 0x6FE4B45: watch_response_cb (kvs_watch.c:217)
==96418==    by 0x6FD6FC3: call_handler (dispatch.c:302)
==96418==    by 0x6FD719E: dispatch_message (dispatch.c:325)
==96418==    by 0x6FD719E: handle_cb (dispatch.c:391)
==96418==    by 0x6FFBD3C: ev_invoke_pending (ev.c:3314)
```

So sorry about this mess!

**Comment by grondo** [comment]
Hm, the above might not be related.. it looks like some possible jansson refcounting issue though.

**Comment by chu11** [comment]
Was rebasing on your branch to prepare my PR #1108.  Noticed you axed ```j_dirent_append()``` in d0125f32a0bad4356250816a6da21e8c2f287155.  I use this in my unit tests in PR #1108.  AFAICT it'll be the last location it can be used, so I could just create a helper function inside the unit tests.  But since I do use it when converting over to jansson, perhaps we can keep ```j_dirent_append()``` for now.  Then I can refactor it out later on?

**Comment by garlick** [comment]
@grondo found the bug - tried to decref a json_t obtained with `json_unpack ("s:o")`.

Just pushed the fix and dropped the proposed lua fix (opened #1112 for future consideration, in case it has merit on its own).  I'll run the PMI test that was failing in a loop during our group meeting and if it doesn't fail in an hour, we're probably past that one.  Whew!  Thanks @grondo!

**Comment by garlick** [comment]
@chu11: ok I'll drop the commit that removes that function if it's more convenient to leave it in for now.

**Comment by garlick** [comment]
I dropped the commit @chu11 wanted dropped, and verified the PMI test has been running in a loop for about 45m with no failures (where before it usually failed within a few minutes).  I think this one's perhaps ready?

**Comment by grondo** [comment]
Yeah, looks good to me!

**Pull request merged**